src/main/java/frc/robot/swerve/ModuleIOReal.java

```diff
@@ -0,0 +1,165 @@
+        angleMotor.getConfigurator().apply(angleMotorConfig);
+
+        driveMotor.setNeutralMode(NeutralModeValue.Brake);
+        driveMotor.setInverted(true);
+        angleMotor.setNeutralMode(NeutralModeValue.Brake);
+        angleMotor.setInverted(true);
+    }
+
+    @Override
+    public void updateInputs(SwerveModuleInputs inputs) {
+        inputs.absolutePosition = encoder.getAbsolutePosition();
+
+        inputs.driveMotorSupplyCurrent = driveMotor.getSupplyCurrent().getValue();
+        inputs.driveMotorStatorCurrent = driveMotor.getStatorCurrent().getValue();
+        driveSupplyChargeUsedCoulomb.update(inputs.driveMotorSupplyCurrent);
+        inputs.driveMotorSupplyCurrentOverTime = driveSupplyChargeUsedCoulomb.get();
+        driveStatorChargeUsedCoulomb.update(inputs.driveMotorStatorCurrent);
+        inputs.driveMotorStatorCurrentOverTime = driveStatorChargeUsedCoulomb.get();
+        inputs.driveMotorPosition = driveMotor.getRotorPosition().getValue();
+        inputs.driveMotorVelocity = getVelocity();
+        inputs.driveMotorVelocitySetpoint = driveMotorVelocitySetpoint;
+
+        inputs.angleMotorSupplyCurrent = angleMotor.getSupplyCurrent().getValue();
+        inputs.angleMotorStatorCurrent = angleMotor.getStatorCurrent().getValue();
+        angleSupplyChargeUsedCoulomb.update(inputs.angleMotorSupplyCurrent);
+        inputs.angleMotorSupplyCurrentOverTime = angleSupplyChargeUsedCoulomb.get();
+        angleStatorChargeUsedCoulomb.update(inputs.angleMotorStatorCurrent);
+        inputs.angleMotorStatorCurrentOverTime = angleStatorChargeUsedCoulomb.get();
+        inputs.angleMotorPosition = angleMotor.getRotorPosition().getValue();
+        inputs.angleMotorVelocity = angleMotor.getVelocity().getValue();
+
+        inputs.angle = getAngle();
+        currentAngle = inputs.angle;
+
+        inputs.angleSetpoint = angleSetpoint;
+
+        inputs.moduleDistance = getModulePosition().distanceMeters;
+        inputs.moduleState = getModuleState();
+    }
+
+    @Override
+    public Rotation2d getAngle() {
+        return new Rotation2d(AngleUtil.normalize(Units.rotationsToRadians(angleMotor.getPosition().getValue())));
+    }
+
+    @Override
+    public void setAngle(Rotation2d angle) {
+        angleSetpoint = angle;
+        Rotation2d error = angle.minus(currentAngle);
+        angleControlRequest.withPosition(angleMotor.getPosition().getValue() + error.getRotations());
+        angleMotor.setControl(angleControlRequest);
+    }
+
+    @Override
+    public double getVelocity() {
+        return driveMotor.getVelocity().getValue();
+    }
+
+    @Override
+    public void setVelocity(double velocity) {
+        var angleError = angleSetpoint.minus(currentAngle);
+        velocity *= angleError.getCos();
+
+        driveMotorVelocitySetpoint = velocity;
+
+        velocityControlRequest.withVelocity(utils.units.Units.metersToRotations(
+                        velocity,
```

Comment:
Looks mixed again -- using distance function with a velocity argument.

Reply:
The calculation for meters to rations is the same as velocity to rps. It was called velocityToRps and you told me to change it to something more general.

Reply:
True, but now that you added a new conversion method, why not use it?

Reply:
What conversion method?

Reply:
You can use `rpsToMetersPerSecond` and invert the result. Could the new units library be helpful here?

Reply:
I'll take your suggestion, the new units library can't convert rotations to distance.

Reply:
Yes, they're the not the same physical dimension. But you could still use `Velocity<Angle>` for the rotational velocity.
I wonder not though why is the conversion needed in the first place? Why not just command the motor with m/s?

Reply:
>But you could still use Velocity<Angle> for the rotational velocity.

Isn't `Rotaion2d` better? Maybe I'm not understanding correctly.

>Why not just command the motor with m/s?

We decided to use rotations, if you want you can try to convince Eitan.

Reply:
> Isn't Rotaion2d better? Maybe I'm not understanding correctly.

It won't matter much in this case. IMO, using `Velocity<Angle>` or `Velcoity<Distance>` will make this a little clearer, because now you know we're measuring a velocity here and not angle/distance. IMO, using meters is better because vendors talk about the speed of swerve modules in terms of "feet per second" and not RPM. Also, that's the measure used in `SwerveModuleState` and eventually in odometry.

Generally, when I see multiple options (in this case, `Rotation2d`/`Velocity<>` and rotations/meters), I look at the following things:
- **Which is clearer?** `Velocity<>` might be clearer, because usually `Rotation2d` is usually not used when measuring speeds.
- **Which makes my code simpler?** `Rotation2d` has the benefit of built-in trigonometric functions, but it's not helpful here. w.r.t rotations/meters, using meters might be easier when looking at other parts of our code.
- **What are my teammates more familiar with**? `Rotation2d` is older, so that's a point to it. However, using new features is generally a good idea in the long run.

@vichik123, would love to hear your opinion.

Reply:
We'll look at it next off-season :)